README

	monocle README:


  * description:
   Monocle is a local network host discovery tool.
   Tested on Linux and FreeBSD 8.
   It can be used in a passive or an active mode.
   The passive mode will just listen for ARP request and reply packets.
   The active mode will send ARP requests to the specified IP address range.
   The result will be a list of IP and MAC address present on the local network.

   * set target
    If you don't define any target, all hosts in the network will be scanned.
    If you want to scan only x.x.x.1 and x.x.x.17 hosts you.
    have to define it such as ./monocle 1,17.
    you can define a host range in this way: ./monocle 1-50.
    it will scan all of the hosts from 1 to 50 (included).
    for instance: in order to scan x.x.x.2/4/6/8/10-50 hosts
    # ./monocle 2,4,6,8,10-50

  * arguments:
   -h, --help                 show usage and exit program.
   -V, --version              show program version and exit.
   -d, --debug                set debug mode.
   -v, --verbose              set verbose mode.
   -t, --timer                show diff time to send and recv packet.
   -H, --no-cols-header       not print header for result scan.
   -i, --device               set network interface.
   -p, --passiv               not send, just listen and catch packet.
   -w, --wait-time            time to the program wait arp response (juste with --passiv).
   -m, --max-pkt              max packet recvied (just with --passiv).
   -s, --stats                show packet stats.
   -e, --send-timeout         wait time for sending next packet.
	                      exemple: 10, wait 10second, m10 wait 10 microsecond.
			      not with --passiv mode.

  * compiling:
    $ ./configure
    $ make
    $ ./monocle --help

NEWS

	monocle NEWS:

  * version 1.0 :    first version


src/mem.c
/*
 *  monocle
 *  src/mem.c
 *
 *  Copyright (C) 2011
 *
 *  Author: Vilmain Nicolas
 *  Contact: null.sim@gmail.com
 *
 *  This file is part of monocle.
 *
 *  monocle is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  monocle is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with monocle.  If not, see <http://www.gnu.org/licenses/>.
 */

#include  "monocle.h"

static void  out_memory (const char *err);

static void
out_memory (const char *err)
{
  fprintf (stderr, "%s:%s: memory exhausted\n",
	   program_name, err);
  exit (EXIT_FAILURE);
}

void *
xcalloc (size_t size)
{
  void *ptr = NULL;

  ptr = malloc (size);
  if (!ptr)
    out_memory ("malloc");
  memset (ptr, 0, size);
  return ptr;
}

void *
xmalloc (size_t size)
{
  void *ptr = NULL;

  ptr = malloc (size);
  if (!ptr)
    out_memory ("malloc");
  return ptr;
}


src/error.c
/*
 *  monocle
 *  src/error.c
 *
 *  Copyright (C) 2011
 *
 *  Author: Vilmain Nicolas
 *  Contact: null.sim@gmail.com
 *
 *  This file is part of monocle.
 *
 *  monocle is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  monocle is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with monocle.  If not, see <http://www.gnu.org/licenses/>.
 */

#include  "monocle.h"

void
error (const char *fmt, ...)
{
  va_list  va;

  va_start (va, fmt);
  if (!debug)
    fprintf (stderr, "%s:", program_name);
  vfprintf (stderr, fmt, va);
  va_end (va);
}


src/error.h
/*
 *  monocle
 *  src/error.h
 *
 *  Copyright (C) 2011
 *
 *  Author: Vilmain Nicolas
 *  Contact: null.sim@gmail.com
 *
 *  This file is part of monocle.
 *
 *  monocle is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  monocle is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with monocle.  If not, see <http://www.gnu.org/licenses/>.
 */

#ifndef __ERROR_H__
#define __ERROR_H__

extern char *program_name;
extern int debug;

#ifdef __GNUC__
# define  __FUNCTION_NAME__  __FUNCTION__
#else
# define  __FUNCTION_NAME__  __func__
#endif /* __GNUC__ */

#define DEBUG(i)					\
  if (debug)						\
    {							\
      fprintf (stderr, "%s:%s:%s:%d:", program_name,	\
	       __FILE__, __FUNCTION_NAME__,		\
	       (__LINE__ - i));				\
    }

void error (const char *fmt, ...);

#endif /* not __ERROR_H__ */


src/ostype.h
/*
 *  monocle
 *  srv/ostype.h
 *
 *  Copyright (C) 2011
 *
 *  Author: Vilmain Nicolas
 *  Contact: null.sim@gmail.com
 *
 *  This file is part of monocle.
 *
 *  monocle is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  monocle is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with monocle.  If not, see <http://www.gnu.org/licenses/>.
 */

#ifndef __OSTYPE_H__
#define __OSTYPE_H__
#define __Linux
#endif /* not __OSTYPE_H__ */


src/socket.c
/*
 *  monocle
 *  src/socket.c
 *
 *  Copyright (C) 2011
 *
 *  Author: Vilmain Nicolas
 *  Contact: null.sim@gmail.com
 *
 *  This file is part of monocle.
 *
 *  monocle is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  monocle is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with monocle.  If not, see <http://www.gnu.org/licenses/>.
 */

#include  "monocle.h"

static int   recv_timeout (int fd, MONOCLE *monocle);
static int   xselect (int fd, fd_set *rd, struct timeval *tv);
#ifdef __BSD
static int    get_bpf (const char *name);
static int    conf_bpf (int fd, const char *device_name);
#endif /* __BSD */
static int    xread (int fd, void *buf, size_t size);

int
get_ethsock (struct ethsock_s *s, int index, const char *name)
{

#ifdef __Linux
  (void) name;
  s->fdsock = socket (AF_PACKET, SOCK_RAW, ETH_ARP);
  if (s->fdsock == -1)
    {
      DEBUG (3);
      error ("socket: %s\n", strerror (errno));
      return -1;
    }
  memset (&s->ll, 0, SOCK_LL_SIZE);
  s->ll.sll_family = AF_PACKET;
  s->ll.sll_protocol = htons (ETH_ARP);
  s->ll.sll_ifindex = index;
  if (bind (s->fdsock, (struct sockaddr *) &s->ll, SOCK_LL_SIZE) == -1)
    {
      DEBUG (2);
      error ("bind: %s\n", strerror (errno));
      close (s->fdsock);
      return -1;
    }

#elif defined(__BSD)
  (void) index;
  s->fdsock = get_bpf (name);
  if (s->fdsock == -1)
    return -1;
#endif /* not __Linux */

  return 0;
}

int
ethsend (struct ethsock_s *s, void *packet, size_t size)
{
  int n_send;

#ifdef __Linux
  n_send = sendto (s->fdsock, packet, size, 0,
		   (struct sockaddr *) &s->ll, SOCK_LL_SIZE);
  if (n_send == -1)
    {
      DEBUG (4);
      error ("send: %s\n", strerror (errno));
    }
#elif defined (__BSD)
  n_send = write (s->fdsock, packet, size);
  if (n_send == -1)
    {
      DEBUG (3);
      error ("write: %s\n", strerror (errno));
    }
#endif /* not __Linux */
  else if ((size_t) n_send != size)
    error ("WARNING: maybe packet send is troncated\n");
  return n_send;
}

int
ethrecv (int fdsock, void *packet, MONOCLE *monocle)
{
#ifdef __Linux
  int ret;
#endif /* __Linux */
  static char st_packet[PACKET_SIZE];
  struct arp_packet_s *pktrecv = NULL;
#ifdef __BSD
  static size_t offset = 0;
  static size_t totsize = 0;
  struct bpf_hdr *bpf = NULL;
#endif /* __BSD */

#ifdef __Linux
  ret = recv_timeout (fdsock, monocle);
  if (ret < 1)
    return ret;
  ret = xread (fdsock, st_packet, PACKET_SIZE);
  if (ret == -1)
    {
      DEBUG (3);
      error ("read: %s\n", strerror (errno));
      return -1;
    }
  pktrecv = (struct arp_packet_s *) st_packet;
#elif defined (__BSD)
  if (offset >= totsize)
    {
      offset = 0;
      totsize = 0;
      totsize = recv_timeout (fdsock, monocle);
      if (totsize < 1)
	return totsize;
      memset (st_packet, 0, PACKET_SIZE);
      totsize = xread (fdsock, st_packet, PACKET_SIZE);
      if ((int) totsize == -1)
	{
	  DEBUG (3);
	  error ("read: %s\n", strerror (errno));
	  return -1;
	}
    }
  bpf = (struct bpf_hdr *) (st_packet + offset);
  pktrecv = (struct arp_packet_s *) (st_packet
				     + offset
				     + bpf->bh_hdrlen);
#endif /* not __Linux */
  monocle->nrecv++;
  if (pktrecv->eth_protocol == htons (ETH_ARP))
    {
      if (monocle->stat)
	monocle->stat->ps_arp++;
#ifdef __Linux
      memcpy (packet, st_packet, ARP_SIZE);
#elif defined (__BSD)
      memcpy (packet, (st_packet + offset + bpf->bh_hdrlen), ARP_SIZE);
      offset += BPF_WORDALIGN(bpf->bh_hdrlen + bpf->bh_caplen);
#endif /* not __Linux */
      return 1;
    }
#ifdef __BSD
  offset += BPF_WORDALIGN(bpf->bh_hdrlen + bpf->bh_caplen);
#endif /* __BSD */
  return 0;
}

static int
recv_timeout (int fd, MONOCLE *monocle)
{
  int val;
  fd_set read_ok; 
  struct timeval  tv;

  FD_ZERO (&read_ok);
  FD_SET (fd, &read_ok);
  if ((monocle->opt & PASSIV))
    {
      if (!monocle->time)
	return xselect (fd + 1, &read_ok, NULL);
      tv.tv_usec = 0;
      val =  monocle->time -
	(time (NULL) - monocle->start_time);
      tv.tv_sec = (val) ? val : 1;
      return xselect (fd + 1, &read_ok, &tv);
    }
  tv.tv_usec = 3900;
  tv.tv_sec = 0;
  return xselect (fd + 1, &read_ok, &tv);
}

static int
xselect (int fd, fd_set *rd, struct timeval *tv)
{
  int  ret;

  ret = select (fd, rd, NULL, NULL, tv);
  if (ret == -1)
    {
      DEBUG (3);
      if (errno != EINTR)
	error ("select: %s\n", strerror (errno));
    }
  return ret;
}

#ifdef __BSD
static int
get_bpf (const char *name)
{
  int   n;
  int   bpf_fd;
  char  bpf_path[15];

  n = 0;
  do
    {
      snprintf (bpf_path, sizeof (bpf_path) - 1,
		"/dev/bpf%d", n);
      bpf_fd = open (bpf_path, O_RDWR);
    }
  while (bpf_fd == -1 && ++n < 10);
  if (bpf_fd == -1)
    {
      error ("cannot find bpf device\n");
      return -1;
    }
  if (conf_bpf (bpf_fd, name))
    {
      close (bpf_fd);
      return -1;
    }
  return bpf_fd;
}

static int
conf_bpf (int fd, const char *device_name)
{
  int bpf_opt;
  struct ifreq  ifr;

  bpf_opt = PACKET_SIZE;
  if (ioctl (fd, BIOCSBLEN, &bpf_opt) == -1)
    {
      DEBUG (2);
      error ("ioctl (BIOCSBLEN): %s\n", strerror (errno));
      return -1;
    }
  strncpy (ifr.ifr_name, device_name, 10);
  if (ioctl (fd, BIOCSETIF, &ifr) == -1)
    {
      DEBUG (2);
      error ("ioctl (BIOCSETIF): %s\n",  strerror (errno));
      return -1;
    }
  bpf_opt = 1;
  if (ioctl (fd, BIOCIMMEDIATE, &bpf_opt) == -1)
    {
      DEBUG (2);
      error ("ioctl (BIOCIMMEDIATE): %s\n", strerror (errno));
      return -1;
    }
  return 0;
}
#endif /* __BSD */

static int
xread (int fd, void *buf, size_t size)
{
  int ret;

  do
    ret = read (fd, buf, size);
  while (ret == -1 && errno == EINTR);
  return ret;
}


src/main.c
/*
 *  monocle
 *  src/main.c
 *
 *  Copyright (C) 2011
 *
 *  Author: Vilmain Nicolas
 *  Contact: null.sim@gmail.com
 *
 *  This file is part of monocle.
 *
 *  monocle is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  monocle is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with monocle.  If not, see <http://www.gnu.org/licenses/>.
 */

#include  "monocle.h"
#include  <getopt.h>

static int     decode_program_options (int argc, char **argv, MONOCLE *monocle);
static int     set_target (int argc, char **argv, MONOCLE *monocle);
static int     parse_ip_args (const char *arg, uint8_t *mask, MONOCLE *monocle);
static void    add_target_in_tab (int bit, uint8_t *mask, MONOCLE *monocle);
static void    free_monocle (MONOCLE *monocle);
static int     monocle_start (MONOCLE *monocle, struct ethsock_s *sock);
static int     monocle_passiv (MONOCLE *monocle, int fdsock);
static int     parse_response (struct arp_packet_s *ar, MONOCLE *monocle);
static void    show_monocle_result (MONOCLE *monocle);
static void    print_target_data (struct target_s *target, const char *offset_mac);
static char *  get_date (time_t timestamp);
static int     xstrtol_positiv_value (const char *str, int *dec);
static int     set_signal_handler (MONOCLE *monocle);
static void    signal_handler (int sig_num);
static void    version (void);
static void    usage (void);

int
main (int argc, char **argv)
{
  int ret;
  MONOCLE monocle;
  struct ethsock_s sock;

  debug = 0;
  gmono = &monocle;   /* just use in signal_handler */
  program_name = argv[0];
  ret = decode_program_options (argc, argv, &monocle);
  if (!ret)
    {
      if (get_ethsock (&sock, monocle.nc.nc_index,
		       monocle.nc.nc_name) == -1)
	{
	  free_monocle (&monocle);
	  return EXIT_FAILURE;
	}
      if ((monocle.opt & PASSIV))
	ret = monocle_passiv (&monocle, sock.fdsock);
      else
	ret = monocle_start (&monocle, &sock);
      close (sock.fdsock);
      show_monocle_result (&monocle);
    }
  free_monocle (&monocle);
  return (!ret) ? EXIT_SUCCESS : EXIT_FAILURE;
}

static int
decode_program_options (int argc, char **argv, MONOCLE *monocle)
{
  char *device_name = NULL;
  char current_opt;
  struct option const long_opt[] =
    {
      {"help",             no_argument,        0,   'h'},
      {"version",          no_argument,        0,   'V'},
      {"debug",            no_argument,        0,   'd'},
      {"verbose",          no_argument,        0,   'v'},
      {"timer",            no_argument,        0,   't'},
      {"send-timeout",     required_argument,  0,   'e'},
      {"no-cols-header",   no_argument,        0,   'H'},
      {"device",           required_argument,  0,   'i'},
      {"passiv",           no_argument,        0,   'p'},
      {"wait-time",        required_argument,  0,   'w'},
      {"max-pkt",          required_argument,  0,   'm'},
      {"stats",            no_argument,        0,   's'},
      {0,                  0,                  0,   0}
    };

  memset (monocle, 0, sizeof (MONOCLE));
  monocle->target = xcalloc (256 * sizeof (struct target_s *));
  do
    {
      current_opt = getopt_long (argc, argv, "hVdpi:w:m:tHsve:", long_opt, NULL);
      switch (current_opt)
	{
	case 'h':
	  usage ();
	  break;
	case 'V':
	  version ();
	  break;
	case 'd':
	  debug = 1;
	  break;
	case 'v':
	  monocle->opt |= VERBOSE;
	  break;
	case 'i':
	  device_name = argv[optind - 1];
	  break;
	case 'p':
	  monocle->opt |= PASSIV;
	  break;
	case 'w':
	  if (xstrtol_positiv_value (optarg, &monocle->time))
	    return -1;
	  break;
	case 'm':
	  if (xstrtol_positiv_value (optarg, &monocle->max_packet))
	    return -1;
	  break;
	case 'e':
	  if (*optarg == 'm')
	    {
	      if (xstrtol_positiv_value ((optarg + 1), &monocle->usec_send))
		return -1;
	    }
	  else
	    {
	      if (xstrtol_positiv_value (optarg, &monocle->sec_send))
		return -1;
	    }
	  break;
	case 't':
	  monocle->opt |= MS_TIME;
	  break;
	case 'H':
	  monocle->opt |= NO_SHOW_HDR;
	  break;
	case 's':
	  monocle->stat = xcalloc (sizeof (struct pktstat_s));
	  break;
	}
    }
  while (current_opt != -1);
  if (get_netconf (&monocle->nc, device_name)
      || set_target (argc, argv, monocle))
    return -1;
  return set_signal_handler (monocle);
}

static int
set_target (int argc, char **argv, MONOCLE *monocle)
{
  int i;
  uint8_t mask[4];
  char ip[17];

  memcpy (mask, &monocle->nc.nc_ipv4, 3);
  if (optind == argc)
    {
      mask[3] = 0;
      printf ("you not set target, target list is ");
      if (inet_ntop (AF_INET, mask, ip, 16))
	printf ("%s-255, try --help for program usage\n", ip);
      else
	printf ("0 at x.x.x.255, try --help for program usage\n");
      for ( i = 0; i < 256; i++)
	add_target_in_tab (i, mask, monocle);
    }
  else
    {
      do
	{
	  if (parse_ip_args (argv[optind], mask, monocle))
	    {
	      error ("invalid argument <%s>\n", argv[optind]);
	      return -1;
	    }
	}
      while (((++optind) - argc));
    }
  return 0;
}

static int
parse_ip_args (const char *arg, uint8_t *mask, MONOCLE *monocle)
{
  int bit;
  int old_bit;
  char delim;

  delim = 0;
  old_bit = 0;
  do
    {
      if (*arg != ',' && *arg != '-')
	{
	  bit = 0;
	  while (*arg && *arg != ',' && *arg != '-')
	    {
	      if (*arg < '0' || *arg > '9')
		return -1;
	      bit *= 10;
	      bit += *arg - '0';
	      arg++;
	    }
	  if (bit > 255)
	    return -1;
	  if (delim == '-')
	    {
	      if (old_bit > bit)
		{
		  for (old_bit--; old_bit >= bit; old_bit--)
		    add_target_in_tab (old_bit, mask, monocle);
		}
	      else
		{
		  for (old_bit++; old_bit <= bit; old_bit++)
		    add_target_in_tab (old_bit, mask, monocle);
		}
	    }
	  else
	    add_target_in_tab (bit, mask, monocle);
	  old_bit = bit;
	}
      delim = *arg;
    }
  while (*arg++);
  return 0;
}

static void
add_target_in_tab (int bit, uint8_t *mask, MONOCLE *monocle)
{
  if (!monocle->target[bit])
    {
      monocle->nhost++;
      monocle->target[bit] = xcalloc (sizeof (struct target_s));
      mask[3] = bit;
      COPY_IP (monocle->target[bit]->ip, mask);
      if (!memcmp (mask, &monocle->nc.nc_ipv4, 4))
	COPY_MAC (monocle->target[bit]->mac, monocle->nc.nc_mac);
    }
}

static void
free_monocle (MONOCLE *monocle)
{
  int i;

  if (monocle->stat)
    free (monocle->stat);
  if (monocle->target)
    {
      for (i = 0; i < 256; i++)
	{
	  if (monocle->target[i])
	    free (monocle->target[i]);
	}
      free (monocle->target);
    }
}

static int
monocle_start (MONOCLE *monocle, struct ethsock_s *sock)
{
  int i;
  int ret;
  int nhost;
  struct arp_packet_s ar_send;
  struct arp_packet_s ar_recv;
  struct timeval buf_time;

  if ((monocle->opt & VERBOSE))
    {
      printf ("%s monocle started with %d target, "
	      "use net device %s, CTRL-C for exit program before the end\n",
	      get_date (time (NULL)), monocle->nhost,
	      monocle->nc.nc_name);
    }
  i = 0;
  nhost = monocle->nhost;
  memset (&ar_send, 0, ARP_SIZE);
  COPY_MAC (ar_send.eth_src, monocle->nc.nc_mac);
  COPY_MAC (ar_send.eth_dst, BROADCAST_ADDR);
  ar_send.eth_protocol = htons (ETH_ARP);
  ar_send.arp_hrd = htons (1);
  ar_send.arp_protocol = htons (ETH_IP);
  ar_send.arp_hln = MAC_SIZE;
  ar_send.arp_pln = IP4_SIZE;
  ar_send.arp_opcode = htons (ARP_OPCODE_REQUEST);
  COPY_MAC (ar_send.arp_mac_src, monocle->nc.nc_mac);
  COPY_IP (ar_send.arp_ip_src, &monocle->nc.nc_ipv4);
  while (nhost && i < 256)
    {
      while (i < 256)
	{
	  if (monocle->target[i])
	    {
	      if (monocle->sec_send)
		sleep (monocle->sec_send);
	      else if (monocle->usec_send)
		usleep (monocle->usec_send);
	      COPY_IP (ar_send.arp_ip_dst, &monocle->target[i]->ip);
	      if (ethsend (sock, &ar_send, ARP_SIZE) == -1)
		return -1;
	      if ((monocle->opt & MS_TIME))
		{
		  gettimeofday (&buf_time, NULL);
		  monocle->target[i]->usec_send = buf_time.tv_usec;
		}
	      i++;
	      break;
	    }
	  i++;
	}
      if (!(--nhost))
	usleep (25000);
      do
	{
	  ret = ethrecv (sock->fdsock, &ar_recv, monocle);
	  if (ret == -1)
	    return -1;
	  else if (ret && (ar_recv.arp_opcode == htons (ARP_OPCODE_REPLY)))
	    {
	      if (monocle->stat)
		monocle->stat->ps_arp_query++;
	      ret = parse_response (&ar_recv, monocle);
	      if (ret > -1 && (monocle->opt & MS_TIME))
		{
		  gettimeofday (&buf_time, NULL);
		  monocle->target[ret]->usec_recv = buf_time.tv_usec;
		}
	    }
	}
      while (ret);
    }
  return 0;
}

static int
monocle_passiv (MONOCLE *monocle, int fdsock)
{
  int ret;
  struct arp_packet_s ar;

  if ((monocle->opt & VERBOSE))
    {
      printf ("%s monocle started with %d target, "
	      "use net device %s, CTRL-C for exit program\n",
	      get_date (time (NULL)), monocle->nhost,
	      monocle->nc.nc_name);
    }
  if (monocle->time)
    monocle->start_time = time (NULL);
  for (;;)
    {
      if ((monocle->time
	   && (time (NULL) - monocle->start_time) > monocle->time)
	  || (monocle->max_packet
	      && monocle->nrecv == monocle->max_packet))
	break;
      ret = ethrecv (fdsock, &ar, monocle);
      if (ret == -1)
	return -1;
      else if (ret && (ar.arp_opcode == htons (ARP_OPCODE_REPLY)))
	{
	  if (monocle->stat)
	    monocle->stat->ps_arp_query++;
	  (void) parse_response (&ar, monocle);
	}
      else if (ret && (ar.arp_opcode == htons (ARP_OPCODE_REQUEST)))
	{
	  if (monocle->stat)
	    monocle->stat->ps_arp_req++;
	  (void) parse_response (&ar, monocle);
	}	
    }
  return 0;
}

static int
parse_response (struct arp_packet_s *ar, MONOCLE *monocle)
{
  int idx;

  idx = ar->arp_ip_src[3];
  if (monocle->target[idx] && !CMP_MAC (monocle->target[idx]->mac,
					NULL_MAC))
    {
      monocle->result++;
      COPY_MAC (monocle->target[idx]->mac, ar->eth_src);
      if ((monocle->opt & VERBOSE))
	{
	  printf ("VERBOSE: find host : ");
	  print_target_data (monocle->target[idx], " ");
	  putchar ('\n');
	  fflush (stdout);
	}
      return idx;
    }
  return -1;
}

static void
show_monocle_result (MONOCLE *monocle)
{
  int i;
  int diff_time;

  if (!monocle->result)
    {
      puts ("no result find");
      return;
    }
  if (!(monocle->opt & NO_SHOW_HDR))
    {
      printf ("IP\r\t\tMAC");
      if ((monocle->opt & MS_TIME) && !(monocle->opt & PASSIV))
	printf ("\r\t\t\t\t\tTIME");
      putchar ('\n');
    }
  for (i = 0; i < 256; i++)
    {
      if (monocle->target[i]
	  && CMP_MAC (monocle->target[i]->mac, NULL_MAC))
	{
	  print_target_data (monocle->target[i], "\r\t\t");
	  if ((monocle->opt & MS_TIME) && !(monocle->opt & PASSIV))
	    {
	      diff_time = (monocle->target[i]->usec_recv
			   - monocle->target[i]->usec_send) / 1000;
	      printf ("\r\t\t\t\t\t%d ms", (diff_time < 1) ? 0 : diff_time);
	    }
	  if (!memcmp (monocle->target[i]->ip, &monocle->nc.nc_ipv4, 4))
	    printf (" (is you)");
	  putchar ('\n');
	}
    }
  if (monocle->stat)
    {
      printf ("packet(s) tot send/recv: %d/%d\n"
	      "packet(s) arp recevied total/request/query: %d/%d/%d\n",
	      (monocle->opt & PASSIV) ? 0 : monocle->nhost, monocle->nrecv, monocle->stat->ps_arp,
	      monocle->stat->ps_arp_req, monocle->stat->ps_arp_query);
    }
}

static void
print_target_data (struct target_s *target, const char *offset_mac)
{
  char ip[17];

  if (inet_ntop (AF_INET, target->ip, ip, 16))
    {
      printf ("%s %s%02x:%02x:%02x:%02x:%02x:%02x",
	      ip, offset_mac, PRINT_MAC_ADDRS (target->mac));
    }
}

static char *
get_date (time_t timestamp)
{
  char *p_date;
  struct tm *tm = NULL;

  tm = localtime (&timestamp);
  if (!tm)
    return "date=unknow";
  strftime (date, DATESIZE - 1, "%G-%m-%d %T", tm);
  p_date = date;
  return p_date;
}

static int
xstrtol_positiv_value (const char *str, int *dec)
{
  char *err = NULL;

  *dec = strtol (str, &err, 0);
  if (*err != '\0')
    {
      error ("%s is not integer value\n", str);
      return -1;
    }
  if (*dec < 1)
    {
      error ("%s is not positiv value\n", str);
      return -1;
    }
  return 0;
}

static int
set_signal_handler (MONOCLE *monocle)
{
  struct sigaction sigact;

  memset (&sigact, 0, sizeof (struct sigaction));
  sigact.sa_handler = &signal_handler;
  if (sigaction (SIGINT, &sigact, NULL) == -1)
    {
      DEBUG (2);
      error ("sigaction(SIGINT): %s\n", strerror (errno));
      if ((monocle->opt & PASSIV) && !monocle->max_packet
	  && !monocle->time)
	return -1;
    }
  return 0;
}

static void
signal_handler (int sig_num)
{
  (void) sig_num;
  puts ("\rsignal SIGINT recevied !");
  show_monocle_result (gmono);
  free_monocle (gmono);
  exit (EXIT_SUCCESS);
}

static void
version (void)
{
  puts ("monocle version 1.0");
  exit (EXIT_SUCCESS);
}

static void
usage (void)
{
  printf ("%s [TARGET] [OPTIONS] ...\n"
	  "If you don't define any target, all hosts in the network will be scanned\n"
	  "If you want to scan only x.x.x.1 and x.x.x.17 hosts you\n"
	  "have to define it such as ./monocle 1,17\n"
	  "you can define a host range in this way: ./monocle 1-50\n"
	  "it will scan all of the hosts from 1 to 50 (included)\n"
	  "for instance: in order to scan x.x.x.2/4/6/8/10-50 hosts -> ./monocle 2,4,6,8,10-50\n"
	  "  -h, --help                 show usage and exit program\n"
	  "  -V, --version              show program version and exit\n" 
	  "  -d, --debug                set debug mode\n"
	  "  -v, --verbose              set verbose mode\n"
	  "  -t, --timer                show diff time to send and recv packet\n"
	  "  -H, --no-cols-header       not print header for result scan\n"
	  "  -i, --device               set network interface\n"
	  "  -p, --passiv               not send, just listen and catch packet\n"
	  "  -w, --wait-time            time to the program wait arp response (just with --passiv)\n"
	  "  -m, --max-pkt              max packet recvied (just with --passiv)\n"
	  "  -s, --stats                show packet stats\n"
	  "  -e, --send-timeout         wait time for sending next packet\n"
	  "                             exemple: 10, wait 10second, m10 wait 10 microsecond\n",
	  program_name);
  exit (EXIT_SUCCESS);
}
